src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCaseEquivalences.cs

```diff
@@ -0,0 +1,165 @@
+        /// <returns><see langword="true"/> if <paramref name="c"/> is involved in case conversion; otherwise, <see langword="false"/></returns>
+        public static bool TryFindCaseEquivalencesForCharWithIBehavior(char c, CultureInfo culture, out ReadOnlySpan<char> equivalences)
+        {
+            if ((c | 0x20) == 'i' || (c | 0x01) == '\u0131')
+            {
+                RegexCaseBehavior mappingBehavior = GetRegexBehavior(culture);
+                equivalences = c switch
+                {
+                    // Invariant mappings
+                    'i' or 'I' when mappingBehavior is RegexCaseBehavior.Invariant => new[] { 'I', 'i' },
+
+                    // Non-Turkish mappings
+                    'i' when mappingBehavior is RegexCaseBehavior.NonTurkish => new[] { 'I', 'i', '\u0130' },
+                    'I' when mappingBehavior is RegexCaseBehavior.NonTurkish => new[] { 'I', 'i', },
+                    '\u0130' when mappingBehavior is RegexCaseBehavior.NonTurkish => new[] { 'i', '\u0130' },
+
+                    // Turkish mappings
+                    'I' or '\u0131' when mappingBehavior is RegexCaseBehavior.Turkish => new[] { 'I', '\u0131' },
+                    'i' or '\u0130' when mappingBehavior is RegexCaseBehavior.Turkish => new[] { 'i', '\u0130' },
+
+                    // Default
+                    _ => null
+                };
+                return equivalences != null;
+            }
+            else
+            {
+                return TryFindCaseEquivalencesForChar(c, out equivalences);
+            }
+        }
+
+        /// <summary>
+        /// Returns which <see cref="RegexCaseBehavior"/> should be used based on the passed in <paramref name="culture"/>.
+        /// </summary>
+        /// <param name="culture">The <see cref="CultureInfo"/> to be used to determine the behavior.</param>
+        /// <returns>The <see cref="RegexCaseBehavior"/> that should be used.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static RegexCaseBehavior GetRegexBehavior(CultureInfo culture)
+        {
+            if (IsCultureInvariant(culture.Name.AsSpan()))
+                return RegexCaseBehavior.Invariant;
+            else if (IsTurkishOrAzerbaycan(culture.Name.AsSpan()))
+                return RegexCaseBehavior.Turkish;
+            else
+                return RegexCaseBehavior.NonTurkish;
+
+            bool IsCultureInvariant(ReadOnlySpan<char> cultureName)
+            {
+                if (culture.Name.Length == 0)
+                    return true;
+                return false;
+            }
+
+            bool IsTurkishOrAzerbaycan(ReadOnlySpan<char> cultureName)
+            {
+                if (cultureName.Length < 2)
+                    return false;
+                else if (cultureName[0] == 't' && cultureName[1] == 'r' && (cultureName.Length == 2 || cultureName[2] == '-'))
```

Comment:
fwiw (not much) this will generate less code:
```c#
            bool IsTurkishOrAzeri(ReadOnlySpan<char> cultureName)

            {
                if (cultureName.Length < 2)
                {
                    return false;
                }
                else if ((cultureName[0] == 't' && cultureName[1] == 'r')||
                         (cultureName[0] == 'a' && cultureName[1] == 'z'))
                {
                    return  (cultureName.Length == 2 || cultureName[2] == '-');
                }
 
               return false;
            }   
```